src/NuGet.Core/NuGet.Packaging/Rules/UpholdBuildConventionRule.cs

```diff
@@ -0,0 +1,114 @@
+
+        public string MessageFormat => AnalysisResources.BuildConventionIsViolatedWarning;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var files = builder.GetFiles().Where(t => t.StartsWith(PackagingConstants.Folders.Build));
+            var packageId = builder.NuspecReader.GetId();
+            var (conventionViolatorsProps, conventionViolatorsTargets) = IdentifyViolators(files, packageId);
+            return GenerateWarnings(conventionViolatorsProps, conventionViolatorsTargets);
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(IDictionary<string, string> conventionViolatorsProps, IDictionary<string, string> conventionViolatorsTargets)
+        {
+            var issues = new List<PackagingLogMessage>();
+            foreach (var props in conventionViolatorsProps)
+            {
+                var warning = string.Format(MessageFormat, _propsExtension, props.Key, props.Value);
+                issues.Add(PackagingLogMessage.CreateWarning(warning, NuGetLogCode.NU5129));
+            }
+
+            foreach (var targets in conventionViolatorsTargets)
+            {
+                var warning = string.Format(MessageFormat, _targetsExtension, targets.Key, targets.Value);
+                issues.Add(PackagingLogMessage.CreateWarning(warning, NuGetLogCode.NU5129));
+            }
+            return issues;
+        }
+
+        internal (IDictionary<string, string>, IDictionary<string, string>) IdentifyViolators(IEnumerable<string> files, string packageId)
+        {
+            var collection = new ContentItemCollection();
+            collection.Load(files.Where(t => PathUtility.GetPathWithDirectorySeparator(t).Count(m => m == Path.DirectorySeparatorChar) > 1));
+
+            var buildItems = ContentExtractor.GetContentForPattern(collection, ManagedCodeConventions.Patterns.MSBuildFiles);
+            var tfms = ContentExtractor.GetGroupFrameworks(buildItems).Select(m => m.GetShortFolderName());
+
+            var conventionViolatorsProps = new Dictionary<string, string>();
+            var conventionViolatorsTargets = new Dictionary<string, string>();
+
+            var correctProps = packageId + _propsExtension;
+            var correctTargets = packageId + _targetsExtension;
+
+            var filesUnderTFM = files.Where(t => t.Count(m => m == '/') > 1);
+            var filesUnderBuild = files.Where(t => t.Count(m => m == '/') == 1);
+
+            if (files.Count() != 0)
+            {
+                if (filesUnderBuild.Count() != 0)
+                {
+                    var hasPropsUnderBuild = filesUnderBuild.Any(t => t.EndsWith(_propsExtension));
+                    var hasTargetsUnderBuild = filesUnderBuild.Any(t => t.EndsWith(_targetsExtension));
+
+                    var correctPropsFilesUnderBuild = filesUnderBuild.Where(t => t.EndsWith(_propsExtension) && t.EndsWith(correctProps));
```

Comment:
I would double check whether the pattern matching for the build folder is case insensitive. 
The reason for this package id being case insensitive.